### PR TITLE
Merge GenDB and SchemaHelper; use GenDB in pclean binary

### PR DESCRIPTION
**Pull request by ThomasColthurst**
This pull request is not 100% ready; there are still a few failing tests in pclean_lib_test and gendb_test.

But I thought I would get your thoughts now to make sure we are on the same page about what this is supposed to do.

Will eventually fix https://github.com/probcomp/hierarchical-irm/issues/207 and https://github.com/probcomp/hierarchical-irm/issues/206

### CODE REVIEW

File: cxx/pclean/pclean.cc [hunk 1]
```diff
@@ -70,77 +70,55 @@ int main(int argc, char** argv) {
...
-  std::cout << "Encoding observations ...\n";
-  T_encoding encoding = calculate_encoding(hirm_schema, encoding_observations);
-
   std::cout << "Incorporating observations ...\n";
-  incorporate_observations(&prng, &hirm, encoding, observations);
+  incorporate_observations(&prng, &gendb, df);
 
   // Run inference
   std::cout << "Running inference ...\n";
-  inference_hirm(&prng, &hirm,
```
Comment (emilyfertig) [inline cxx/pclean/pclean.cc:120]: Can we still run HIRM with a PClean-like schema, i.e. Model 6? I think that's something we should try to keep (maybe with a bool flag for HIRM vs full GenDB, and calling incorporate_observations, inference_hirm etc on gendb.hirm)

**Review by emilyfertig — COMMENTED** [review]
Overall this LGTM. (If we're using pclean.cc to launch GenDB, we should probably move it up a directory and rename it, not necessarily in this PR).

**Comment by ThomasColthurst** [comment]
Tests pass now.

File: cxx/pclean/pclean.cc (hunk 1 above)
Comment (ThomasColthurst) [reply cxx/pclean/pclean.cc:120]: So not only can we do this, it is currently the only thing we can do -- all inference.cc::inference_gendb does right now is directly call inference_hirm.

But I agree that as we add the ability to inference_gendb to transition entity assignments, we should add flags to keep Model 5 & 6 behavior.

**Comment by ThomasColthurst** [comment]
My preference, by the way, would be to move gendb.* down into the pclean directory rather than moving pclean.cc up a directory.

File: cxx/gendb.cc [hunk 2]
```diff
@@ -60,6 +60,9 @@ void GenDB::incorporate(
     // Incorporate the items/value into the query relation.
     incorporate_query_relation(prng, query_rel, items, val);
   }
+
+  // Add to the record_class's CRP.
```
Comment (emilyfertig) [inline cxx/gendb.cc:64]: I don't think we should have a CRP for the Record class. What is the problem, and is there another way to fix it?

File: cxx/pclean/pclean.cc (hunk 1 above)
Comment (emilyfertig) [reply cxx/pclean/pclean.cc:120]: Sounds good. Could you add those, as Model 7 inference comes together?

**Comment by emilyfertig** [comment]
Since GenDB is a combination of HIRM and PClean, maybe we should do something like

cxx/
  pclean/
  hirm/
    relation.hh
    domain.hh
    ...
  distributions/
  ...
  gendb.hh
  gendb_main.cc <- renamed from pclean.cc

In any case I think we can do better than the current structure, I'll file an issue to figure it out.

File: cxx/gendb.hh [hunk 3]
```diff
@@ -125,4 +124,34 @@ class GenDB {
   // Disable copying.
   GenDB& operator=(const GenDB&) = delete;
   GenDB(const GenDB&) = delete;
-};
\ No newline at end of file
+
+  // The rest of these methods are conceptually private, but actually
+  // public for testing.
+
+  void compute_domains_cache();
```
[the remark this answers is not in the document]
Comment (ThomasColthurst) [reply cxx/gendb.hh:123]: Done.

File: cxx/gendb.cc (hunk 2 above)
Comment (ThomasColthurst) [reply cxx/gendb.cc:64]: Well, the specific problem was that I assumed that there would be a CRP for the record class, and I wrote a test under that assumption, and then I added these lines to make that test pass.

But I also use the record class's CRP in pclean_lib.cc::make_pclean_sample to create a class_item for each new row.  I'm open to suggestions for better ways to implement that, but I believe that sampling from the record class's CRP is what the model spec says to do.

Comment (emilyfertig) [reply cxx/gendb.cc:64]: For make_pclean_sample, I think we should just use a counter for class_item, or have the function take a vector of unique row IDs. I don't think the spec says to sample from the record class's CRP -- there's a one-to-one correspondence between observations and record class entities, and sampling from a record class CRP would result in multiple observations of a single record entity.

File: cxx/pclean/pclean.cc (hunk 1 above)
Comment (ThomasColthurst) [reply cxx/pclean/pclean.cc:120]: Sure.

File: cxx/gendb.cc (hunk 2 above)
Comment (ThomasColthurst) [reply cxx/gendb.cc]: Done.  When I was talking about the spec, I was talking about the generative model described on page 7 under "PClean entity model".

**Comment by ThomasColthurst** [comment]
Two of the integration tests are failing with crashes during inference.

File: cxx/gendb.hh [hunk 4]
```diff
@@ -125,4 +110,64 @@ class GenDB {
...
+  HIRM* hirm;  // Owned by the GenDB instance.
+
+  // Map keys are class names. Values are CRPs for latent entities, where the
+  // "tables" are entity IDs and the "customers" are unique identifiers of
+  // observations of that class.
+  std::map<std::string, CRP> domain_crps;
+
+  bool only_final_emissions;
+  bool record_class_is_clean;
+  std::map<std::string, std::vector<std::string>> domains;
```
Comment (emilyfertig) [inline cxx/gendb.hh:160]: Could you add a comment for "domains"?

File: cxx/pclean/pclean_lib.cc [hunk 5]
```diff
@@ -40,89 +35,48 @@ T_observations translate_observations(
...
-      &annotated_domains_for_relations,
-    std::mt19937* prng) {
+DataFrame make_pclean_samples(int num_samples, int start_row, GenDB *gendb,
+                              std::mt19937* prng) {
   DataFrame df;
   for (int i = 0; i < num_samples; i++) {
      std::map<std::string, std::string> query_values;
-     WIP_make_pclean_sample(hirm, schema, annotated_domains_for_relations,
-                        prng, &query_values);
+     make_pclean_sample(prng, gendb, start_row + i, &query_values);
```
Comment (emilyfertig) [inline cxx/pclean/pclean_lib.cc:74]: Maybe add an assertion that start_row isn't already in the record class, or add a comment that explains the assumption (that entity IDs greater than or equal to start_row aren't already in the record class).

**Review by emilyfertig — APPROVED** [review]
Thanks!

File: cxx/pclean/pclean_lib.cc (hunk 5 above)
Comment (ThomasColthurst) [reply cxx/pclean/pclean_lib.cc:78]: Added comment.

**Comment by ThomasColthurst** [comment]
> Thanks!

Integration tests now pass, post merge with https://github.com/probcomp/hierarchical-irm/pull/215 and using new_rows_have_unique_entities=true in pclean_lib::incorporate_observations.

**Pull request merged**